### PR TITLE
[small change] Use selective log-softmax to reduce peak vram consumption

### PR DESCRIPTION
**Pull request by tyler-romero**
When training language models, we often need to convert logits (raw model outputs) into log probabilities. The standard approach uses log_softmax which requires computing probabilities for every token in the vocabulary at every position in the sequence. For large vocabulary sizes, this can consume significant VRAM.

For example, with a modest vocabulary size of 32768, sequence length of 1024, and batch size of 16, computing log_softmax naively can consume 2.1GB of VRAM. And that is in addition to the 2.1GB required to hold the logits in the first place. However, in many cases, we only need the log probabilities for specific tokens - usually the ones that were actually generated or appear in the training data.

We can optimize this by:
1. First gathering just the logits for the tokens we care about
2. Computing the softmax denominator (logsumexp) over the full logit distribution
3. Subtracting the denominator from our gathered logits to get the final log probabilities

In this example setting, vram usage is reduced from 4295MB to 2282MB (2147MB of which is required to store the logits in the first place).

### CODE REVIEW

**Comment by tyler-romero** [comment]
Benchmark Script (Thanks [Quentin Gallouédec](https://github.com/qgallouedec))

```python
import time
import torch

def naive_method(logits, input_ids):
    log_probs = logits.log_softmax(dim=-1)
    return torch.gather(log_probs, dim=-1, index=input_ids.unsqueeze(-1)).squeeze(-1)

def method_1(logits, input_ids):  # compute log_softmax in a loop to reduce peak memory
    per_token_logps = []
    for logits_row, input_ids_row in zip(logits, input_ids):
        log_probs = logits_row.log_softmax(dim=-1)
        token_log_prob = torch.gather(log_probs, dim=-1, index=input_ids_row.unsqueeze(-1)).squeeze(-1)
        per_token_logps.append(token_log_prob)
    return torch.stack(per_token_logps)

def method_2(logits, input_ids):  # avoid materializing unneeded log_probs to reduce peak memory
    token_logits = torch.gather(logits, dim=-1, index=input_ids.unsqueeze(-1)).squeeze(-1)
    logsumexp_values = torch.logsumexp(logits, dim=-1)
    token_log_probs = token_logits - logsumexp_values  # log_softmax(logits) = logits - log(sum(exp(logits)))
    return token_log_probs

def method_3(logits, input_ids):  # combine methods 1 and 2
    per_token_logps = []
    for logits_row, input_ids_row in zip(logits, input_ids):
        token_logits = torch.gather(logits_row, dim=-1, index=input_ids_row.unsqueeze(-1)).squeeze(-1)
        token_log_prob = token_logits - torch.logsumexp(logits_row, dim=-1)
        per_token_logps.append(token_log_prob)
    return torch.stack(per_token_logps)

def efficient_method(logits, input_ids):  # pull everything out of the loop except logsumexp
    token_logits = torch.gather(logits, dim=-1, index=input_ids.unsqueeze(-1)).squeeze(-1)
    logsumexp_values = torch.stack([torch.logsumexp(l, dim=-1) for l in logits])
    token_log_probs = token_logits - logsumexp_values
    return token_log_probs

def measure_memory_and_time(func, logits, input_ids):
    torch.cuda.reset_peak_memory_stats()
    start_time = time.perf_counter()
    result = func(logits, input_ids)
    end_time = time.perf_counter()
    mem_peak = torch.cuda.max_memory_allocated()
    return result, end_time - start_time, mem_peak

# Simulated data
torch.manual_seed(42)
vocab_size = 32768
seq_len = 1024
batch_size = 16

device = "cuda" if torch.cuda.is_available() else "cpu"
logits = torch.randn(batch_size, seq_len, vocab_size, device=device, dtype=torch.float32)
input_ids = torch.randint(0, vocab_size, (batch_size, seq_len), device=device)
logit_mem = torch.cuda.max_memory_allocated()

# Run all methods
naive_result, naive_time, naive_mem = measure_memory_and_time(naive_method, logits, input_ids)
method1_result, method1_time, method1_mem = measure_memory_and_time(method_1, logits, input_ids)
method2_result, method2_time, method2_mem = measure_memory_and_time(method_2, logits, input_ids)
method3_result, method3_time, method3_mem = measure_memory_and_time(method_3, logits, input_ids)
efficient_result, efficient_time, efficient_mem = measure_memory_and_time(efficient_method, logits, input_ids)

# Check equivalence
print("Max absolute difference (naive and 1):", (naive_result - method1_result).abs().max().item())
print("Max absolute difference (naive and 2):", (naive_result - method2_result).abs().max().item())
print("Max absolute difference (naive and 3):", (naive_result - method3_result).abs().max().item())
print("Max absolute difference (naive and efficient):", (naive_result - efficient_result).abs().max().item())
print("Memory consumed by logits: {:.2f} MB".format(logit_mem / 1e6))
print("Naive method time:      {:.6f} sec, Memory peak: {:.2f} MB".format(naive_time, naive_mem / 1e6))
print("Method 1 time:          {:.6f} sec, Memory peak: {:.2f} MB".format(method1_time, method1_mem / 1e6))
print("Method 2 time:          {:.6f} sec, Memory peak: {:.2f} MB".format(method2_time, method2_mem / 1e6))
print("Method 3 time:          {:.6f} sec, Memory peak: {:.2f} MB".format(method3_time, method3_mem / 1e6))
print("Efficient method time:  {:.6f} sec, Memory peak: {:.2f} MB".format(efficient_time, efficient_mem / 1e6))

# Results:
# > Max absolute difference (naive and 1): 0.0
# > Max absolute difference (naive and 2): 1.9073486328125e-06
# > Max absolute difference (naive and 3): 1.9073486328125e-06
# > Max absolute difference (naive and efficient): 1.9073486328125e-06
# > Memory consumed by logits: 2147.61 MB
# > Naive method time:      0.036307 sec, Memory peak: 4295.16 MB
# > Method 1 time:          0.012156 sec, Memory peak: 2416.18 MB
# > Method 2 time:          0.134651 sec, Memory peak: 4295.43 MB
# > Method 3 time:          0.001496 sec, Memory peak: 2282.10 MB
# > Efficient method time:  0.000918 sec, Memory peak: 2282.23 MB
```

In this benchmark setting, peak VRAM usage for this operation was reduced by 47% (from 4295MB to 2282MB). And most of the memory consumed now is due to the size of the input logits (2147MB). The proposed method is faster as well.

In practice the memory consumed by a naive log_softmax can be ~10% of the memory required for the learning phase of GPRO.

**Comment by hijkzzz** [comment]
The excellent MR, I will come by to review when I have time. This can also be used in PPO, right?

**Comment by tyler-romero** [comment]
Yes, can work in PPO as well!

**Comment by wangzhen0518** [comment]
Why method 2 is much slower than efficient method? The only difference between them is that method 2 calculates the logexpsum all at once using tensors, while the efficient method computes the logexpsum sample by sample using a loop. Given that method 2 utilizes tensor parallelism, shouldn't it be faster than the efficient method? Is it related to gpu cache? It may be also the reason why method 2 is slower than method1.

**Comment by HSYZhang** [comment]
Why does method2 perform so much worse than efficient_method? Is it because the underlying operator implementation of torch.logsumexp has issues with the above data scale?

**Comment by HSYZhang** [comment]
> Why does method2 perform so much worse than efficient_method? Is it because the underlying operator implementation of torch.logsumexp has issues with the above data scale?

Even the GPU memory usage hasn't been reduced at all.
